juju/model.py

```diff
@@ -116,29 +132,35 @@ class Relation:
     def __init__(self, relation_name, relation_id, local_unit, backend, cache):
         self.relation_name = relation_name
         self.relation_id = relation_id
-        self._apps = set()
-        self._units = []
+        self.apps = set()
+        self.units = set()
         for unit_name in backend.relation_list(relation_id):
             unit = cache.get(Unit, unit_name)
-            self._units.append(unit)
-            self._apps.add(unit.app)
+            self.units.add(unit)
+            self.apps.add(unit.app)
         self.data = RelationData(relation_name, relation_id, local_unit, self.units, backend)
 
-    @property
-    def apps(self):
-        return self._apps
 
-    @property
-    def units(self):
-        return self._units
+class DeadRelation(Relation):
+    def __init__(self, relation_name, relation_id):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self.apps = set()
+        self.units = set()
+        self.data = RelationData(relation_name, relation_id, None, None, None)
 
 
 class RelationData(Mapping):
     def __init__(self, relation_name, relation_id, local_unit, remote_units, backend):
         self.relation_name = relation_name
         self.relation_id = relation_id
-        self._data = {local_unit: RelationUnitData(relation_id, local_unit, True, backend)}
-        self._data.update({unit: RelationUnitData(relation_id, unit, False, backend) for unit in remote_units})
+        if local_unit:
+            self._data = {local_unit: RelationUnitData(relation_id, local_unit, True, backend)}
+            self._data.update({unit: RelationUnitData(relation_id, unit, False, backend) for unit in remote_units})
+        else:
+            # If we don't have even a local unit, then we're dealing with a dead relation;
+            # and dead relations tell no tales.
```

Comment:
I'm curious about how this will pan out in practice once we start exercising it with more realistic examples. Let's keep our eye on it.